docs/metrics-data-model.md

```diff
@@ -0,0 +1,177 @@
+# k6 Time Series
+
+We would migrate to a time series data model for getting the benefits of aggregated data so just allocating samples one time and pass around the aggregated time series value (or values - e.g. in case of Trend). The Outputs will switch from fetching the buffered samples to fetching a snapshot of the time series (the same concept adopted by Prometheus' scraping operation).
+
+## Data model
+
+#### Metrics and Series
+
+* Metric (name, type)
+* TimeSeries (metric, tagset)
+* Sample (metric, tagset, timestamp, value)
+* TimeSeries Sample (timeseries ref, timestamp, value)
+
+```go
+// This is the current k6 Metric polished from the Thresholds stuff.
+type Metric struct {
+	Name     string     `json:"name"`
+	Type     MetricType `json:"type"`
+	Contains ValueType  `json:"contains"`
+}
+
+type TimeSeries struct {
+    ID uin64 // hash(metric+tags)
+	MetricName string
+	Tags   TagSet
+
+	Sink Sink // Counter, Guage, Rate, Trend (SparseHistogram in the future)
+}
+
+type Sink interface {
+    AddPoint(v float64)
+}
+
+type TSample struct { 
+	TimeSeriesID string
+	Timestamp      uint64
+	Value          float64
+}
+
+type Sample struct {
+        MetricName string
+        Timestamp uint64
+        Tags TagSet
+        Value float64
+}
+```
+
+#### Open questions:
+
+* `TimeseriesSample` knows the time series concept and it has the reference to a time series. The Sample is raw data and there isn't yet an identified/matched time series to reference. In this way, the collecting phase doesn't  require knowledge from a time series database. It could speed up the  implementation and introduction of the first stage, where hopefully we don't touch the sample generation (e.g. all the `k6/js/modules`).
```

Comment:
I think this might not be the best way to go about it :thinking: It would force the calculation of hashes to run in just a single thread. If that was done as soon as a metric is measured in the `js/` package (or elsewhere), that relatively CPU intensive task will be spread apart on multiple cores... :thinking: 

However, if the association with a `TimeSeries` is done there, it will require locking, whereas if it was done centrally from a single thread, it will not require any locking, just a map lookup (or insertion, if the time series didn't exist).

WDYT?

Reply:
I see the central solution as a more straightforward step for introducing this new model in the core. We can extend the solution with the distributed part in the next step.

I collapsed the two structures in the current Sample just adding a TimeSeries pointer that I expect the metrics engine or the ingester will resolve. In this way, we can resolve the submetrics/contains issue in the first step, and in a later PR, we can drop the Tags field from the Sample structure by distributing the time series generation to the components on the edge.

Do you see limits for it?

Reply:
> @na-- I see the central solution as a more straightforward step for introducing this new model in the core. We can extend the solution with the distributed part in the next step.

Hmm... maybe? :sweat_smile: isn't the "distributed" part of the time series determination just calculating the hash when the metric sample is assembled? :thinking: That seems like it would be easier to get done at the start, rather than refactoring it later? If it was just dependent on the tags, it could have probably been done by just modifying things [here](https://github.com/grafana/k6/blob/556747fbe585c981507904d634e705c68891e663/metrics/sample.go#L292-L318), but since it also depends on the metric name, it's going to need some more refactoring :disappointed: